### PR TITLE
Define _LIBCPP_HAS_NO_THREADS when building non-threads version of li…

### PR DESCRIPTION
**Pull request by sbc100**
…bcxx

### CODE REVIEW

File: system/lib/libcxx/include/__config [hunk 1]
```diff
@@ -10,6 +10,18 @@
 #ifndef _LIBCPP_CONFIG
 #define _LIBCPP_CONFIG
 
+// XXX Emscripten: begin site config
+#ifdef __EMSCRIPTEN__
+#define _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS
+// Normally macros like _LIBCPP_HAS_NO_THREADS are fixed at build time,
+// but we to have this header work for both the multi-threaded and
+// single-threaded version of libc++.
+#ifndef _REENTRANT
```
Comment (kripken) [inline system/lib/libcxx/include/__config:19]: what is `_REENTRANT`?

can we not use the normal way to detect atomics in wasm?

File: system/lib/libcxx/include/__config [hunk 2]
```diff
@@ -1160,10 +1172,6 @@ _LIBCPP_FUNC_VIS extern "C" void __sanitizer_annotate_contiguous_container(
        _LIBCPP_HAS_NO_THREADS is defined.
 #endif
 
-#if defined(__STDCPP_THREADS__) && defined(_LIBCPP_HAS_NO_THREADS)
-#error _LIBCPP_HAS_NO_THREADS cannot be set when __STDCPP_THREADS__ is set.
-#endif
```
Comment (kripken) [inline system/lib/libcxx/include/__config:1165]: why remove this? if it's a local change, perhaps `#if 0` with a comment?

Comment (sbc100) [reply system/lib/libcxx/include/__config:1165]: Sorry I forgot to mention, this was removed upstream already in https://reviews.llvm.org/D92349

File: system/lib/libcxx/include/__config (hunk 1 above)
Comment (sbc100) [reply system/lib/libcxx/include/__config:19]: What is the normal way?    The _REENTRANT  macro is defined by clang when building with `-pthread`.

Comment (kripken) [reply system/lib/libcxx/include/__config:19]: We use `__EMSCRIPTEN_PTHREADS__` in other places to check for this.

I wasn't aware of `_REENTRANT`. Not opposed to it, but if we can be consistent that sounds better?

Comment (sbc100) [reply system/lib/libcxx/include/__config]: Done

**Review by kripken — APPROVED** [review]

**Comment by sbc100** [comment]
So I think I may  abandon this change since it has the effect making include of `<atomic.h>` not work without `USE_PTHREADS`. 

This means we can't compile code that uses atomics and have them lowered automatically away (at least not without USE_PTHEADS.    WDYT @tlively @kripken ... is it OK to force folks to compile with `USE_PTHEADS` to use any `<atomic>`headers?

**Comment by tlively** [comment]
> is it OK to force folks to compile with USE_PTHEADS to use any <atomic>headers?

I would rather allow folks to use <atomic> for both pthreads and non-pthreads builds, but I guess I don't know whether our users would care. Do I understand correctly that in upstream clang/libc++ this means that you can't use <atomic> if pthreads is disabled as well? I would be surprised by that, but that would make me more ok with having the same restriction in Emscirpten.

**Comment by sbc100** [comment]
> > is it OK to force folks to compile with USE_PTHEADS to use any headers?
> 
> I would rather allow folks to use for both pthreads and non-pthreads builds, but I guess I don't know whether our users would care. Do I understand correctly that in upstream clang/libc++ this means that you can't use if pthreads is disabled as well? I would be surprised by that, but that would make me more ok with having the same restriction in Emscirpten.

Yes, its not possible to even `#include` the `<atomic>` header if this macro is defined.  Perhaps we should suggest that upstream allow this configuration?

**Comment by tlively** [comment]
Hmm, maybe abandoning this change makes sense, then. The error message that <atomic> gives when `_LIBCPP_HAS_NO_THREADS` is defined is "<atomic> is not supported on this single threaded system." This implies that `_LIBCPP_HAS_NO_THREADS` should only be defined when the entire system is single-threaded. But Emscripten is not a single-threaded system, it just supports single-threaded builds. If we didn't support threads at all, I would have no problem with disallowing the use of <atomic>.

**Pull request closed without merging**